notebooks/generative/fashionmnist_vae.livemd

```diff
@@ -0,0 +1,526 @@
+    names: [:images, :channels, :height, :width]
+  )
+
+Kino.animate(round(5000 / num_steps), 0, fn i ->
+  if i > num_steps do
+    :halt
+  else
+    {:cont, Data.image_to_kino(reconstructed_images[i]), i + 1}
+  end
+end)
+```
+
+Cool! We have interpolation! But did you notice that some of the intermediate frames don't look fashionable at all? Autoencoders don't generally return good results for random vectors in their latent space. That's where a VAE can help.
+
+<!-- livebook:{"branch_parent_index":1} -->
+
+## Making it variational
+
+In a VAE, instead of outputting a latent vector, our encoder will output a distribution. Essentially this means instead of 10 outputs we'll have 20. 10 of them will represent the mean and 10 will represent the log of the variance of the latent. We'll have to sample from this distribution to get our latent vector. Finally, we'll have to modify our loss function to also compute the KL Divergence between the latent distribution and a standard normal distribution (this acts as a regularizer of the latent space).
+
+We'll start by defining our model:
+
+```elixir
+defmodule Vae do
+  import Nx.Defn
+
+  @latent_features 10
+
+  defp sampling_layer(%Axon{} = input, _opts \\ []) do
+    Axon.layer(&sampling_layer_impl/2, [input], name: "sampling_layer", op_name: :sample)
+  end
+
+  defnp sampling_layer_impl(x, _opts \\ []) do
+    mu = x[[0..-1//1, 0, 0..-1//1]]
+    log_var = x[[0..-1//1, 1, 0..-1//1]]
+    std_dev = Nx.exp(0.5 * log_var)
+    eps = Nx.random_normal(std_dev)
+    sample = mu + std_dev * eps
+    Nx.stack([sample, mu, std_dev], axis: 1)
+  end
+
+  defp encoder_partial() do
+    Axon.input("image", shape: {nil, 1, 28, 28})
+    # This is now 28*28*1 = 784
+    |> Axon.flatten()
+    # The encoder
+    |> Axon.dense(256, activation: :relu, name: "encoder_layer_1")
+    |> Axon.dense(128, activation: :relu, name: "encoder_layer_2")
+    |> Axon.dense(64, activation: :relu, name: "encoder_layer_3")
+    # Bottleneck layer
+    |> Axon.dense(@latent_features * 2, name: "bottleneck_layer")
+    # Split up the mu and logvar
+    |> Axon.reshape({2, @latent_features})
+    |> sampling_layer()
+  end
+
+  def encoder() do
+    encoder_partial()
+    # Grab only the sample (ie. the sampled latent)
+    |> Axon.nx(fn x -> x[[0..-1//1, 0]] end)
+  end
+
+  def decoder(input_latent) do
+    input_latent
+    |> Axon.dense(64, activation: :relu, name: "decoder_layer_1")
+    |> Axon.dense(128, activation: :relu, name: "decoder_layer_2")
+    |> Axon.dense(256, activation: :relu, name: "decoder_layer_3")
+    |> Axon.dense(784, activation: :sigmoid, name: "decoder_layer_4")
+    |> CustomLayer.scaling_layer()
+    # Turn it back into a 28x28 single channel image
+    |> Axon.reshape({1, 28, 28})
+  end
+
+  def autoencoder() do
+    encoder_partial = encoder_partial()
+    encoder = encoder()
+    autoencoder = decoder(encoder)
+    Axon.container(%{mu_sigma: encoder_partial, reconstruction: autoencoder})
```

Comment:
Not sure if there was a better way to do this...

Reply:
This is pretty much exactly how I would do it too :)